Simple stream rewrite (Breaking)

trs opened the pull request:
This breaking change revisits how a stream should behave. Instead of interop-ing with native streams, this introduces a "new" type of stream behaviour. PureStreams have no event, propagate errors to piped destinations, and end when an error occurs.

Transformations are performed by piping into a transformer stream:
```js
stream.pipe(map((num) => num + 1);
```

Stream values can be inspected by calling `each`. (Similar to `.on('data', …)`)
```js
stream.each((value) => console.log(value));
```

Streams are consumed by calling `done` on the stream. `done` takes a callback with an `error` arg if one occured. (Similar to `.on('end', …)`)
```js
stream.done((err) => console.log(err));
```

trs committed:
feat: rewrite to use custom stream class

A custom stream class allows us to implement a simpler object stream.
This stream propagates errors on pipe and ends on errors.
No more listen events, instead values are consumed with `.done`

Check the next commit for docs.

BREAKING CHANGE: no longer interop with native streams

trs committed:
docs: update docs to reflect changes

trs committed:
refactor: move internal options

trs committed:
docs: update method jsdocs

matt-forster reviewed (commented):
This looks well designed. I like how the types have been simplified, which makes it easier to reason about. I would try to match some of the existing object APIs to allow for quick understanding, even implemented as aliases.

trs committed:
feat: allow done to not consume stream

Passing `false` to the `consume` option in `.done()` allows multiple
actions when a stream ends, without actually starting to consume the
stream.

This is useful for side effects of a function that return a stream.

trs committed:
test: remove dupliate test

trs committed:
feat: split take into take and slice

Slice works like array slice.
Take stops after x number of items.

trs committed:
docs: clarify method usage

trs committed:
refactor: return bool on write

trs committed:
chore: move to autovance eslint

matt-forster reviewed (approved):
LGTM 🍃

trs committed:
refactor: simplify passthrough

trs committed:
docs: fix formatting
